### ISSUE TITLE
VS 2015

### ISSUE THREAD

**Issue by randomascii** [google/UIforETW#58]
Should UIforETW be upgraded to VS 2015? The free VS 2015 community edition can build UIforETW and the VC++ 2015 compiler has additional C99 and C++ features that may make development easier, such as %zd, constexpr, etc.

**Comment by mwinterb** [comment]
One minor "for free" gain is that upgrading to VS 2015 will fix ElapsedTimer, as steady_clock is finally steady. It will also further blur any differences between ElapsedTimer and QPCElapsedTimer as steady_clock now uses QueryPerformanceCounter.

**Comment by mwinterb** [comment]
Also, /Zc:strictStrings can be enabled in debug mode as well as release, as the debug operator new overloads issue has been addressed.

**Comment by randomascii** [comment]
Thanks for the information. @ariccio - any thoughts?

**Comment by ariccio** [comment]
I really like the idea of upgrading, but haven't yet gotten my hands on a proper version of VS2015... so I wouldn't be able to build UIforETW if we use any VS2015-only features >:(

**Comment by randomascii** [comment]
I used the free VS 2015 community edition. Would that work for you?

If we upgraded then I would want to start using new language features immediately, but I don't want to leave you behind, so I'd rather not upgrade yet if it will block you.

**Comment by ariccio** [comment]
> I used the free VS 2015 community edition. Would that work for you?

Hold up... Microsoft *DOES* include MFC in the free edition? Have I been laboring under the delusion that *VS2015 community edition does not include MFC* for the whole time? If the community edition includes MFC, then there's absolutely no reason to hold back, and I'll enthusiastically upgrade.



Oof, my stupidity truly has no bound.

**Comment by ariccio** [comment]
Ahh, I must've read "doesn't support *multibyte* MFC" as "doesn't support MFC". Since nobody sane uses multibyte MFC, it's a moot point.

**Comment by randomascii** [comment]
I think that the express editions of Microsoft for VS 2013 and earlier did not come with MFC, or at least they were less powerful and more restrictive. The VS 2015 community edition seems remarkably capable.

I'll upgrade when I get a chance, and probably add something to the CONTRIBUTING page so that potential contributors know that 'free' is an option.

**Comment by rpavlik** [comment]
The express editions do not have restrictions on usage eligibility, but lack MFC/ATL and whichever the non deprecated one of addons or extensions is. Community edition is essentially the same feature set as professional, but there are license restrictions on who can use it and for what. (They are in fact parallel installable, so you can have express for your non-community-qualified needs, and thankfully Microsoft did actually release a 2015 express edition.

**Comment by randomascii** [comment]
I seems that the restrictions on the community edition are quite light - you can (IIRC) ship small commercial products with it, so that's cool.

**Comment by ariccio** [comment]
Alright, I can build with VS2015, with a few ~~exceptions~~ errors:

```
uiforetw\uiforetw\uiforetwdlg.cpp(331): error C2220: warning treated as error - no 'object' file generated
uiforetw\uiforetw\uiforetwdlg.cpp(331): warning C4311: 'reinterpret_cast': pointer truncation from 'HINSTANCE' to 'int'
uiforetw\uiforetw\uiforetwdlg.cpp(331): warning C4302: 'reinterpret_cast': truncation from 'HINSTANCE' to 'int'
uiforetw\uiforetw\uiforetwdlg.cpp(349): warning C4311: 'reinterpret_cast': pointer truncation from 'HINSTANCE' to 'int'
uiforetw\uiforetw\uiforetwdlg.cpp(349): warning C4302: 'reinterpret_cast': truncation from 'HINSTANCE' to 'int'
```

...which corresponds to [line 333 in master](https://github.com/google/UIforETW/blob/57a5a9a3e6ea367bf24eb8db942104dbc4e24568/UIforETW/UIforETWDlg.cpp#L333).


It [looks like you encountered exactly those warnings in Chromium](https://code.google.com/p/chromium/issues/detail?id=440500#c76), which is one of exactly 8 Google search results for `"warning C4311: 'reinterpret_cast': pointer truncation from 'HINSTANCE' to 'int'"`.

**Comment by ariccio** [comment]
Huh... so I switched from the [`StrictRules` ruleset](https://github.com/google/UIforETW/blob/master/UIforETW/StrictRules.ruleset) to the VS2015 built in "Microsoft All Rules" ruleset, and I got a couple new "buffer might not be zero-terminated" warnings:

![uietw_nullterm_warning](https://cloud.githubusercontent.com/assets/2142308/10595564/c99d64d4-76a7-11e5-8de1-3726e7f16f6f.PNG)

**Comment by ariccio** [comment]
(this build is of my PR, so there's a warning in utility.cpp that you may not see)

The warnings are for `swprintf_s` and `_vsnwprintf_s`, which don't seem to actually guarantee null-termination on failure. ["Security Features in the CRT"](https://msdn.microsoft.com/en-us/library/8ef0s5kh.aspx) says:

- **Null termination.** Some functions that left potentially non-terminated strings have secure 
versions which ensure that strings are properly null terminated.

Note that it says *some*, but not specifically what functions have that guarantee.
So these are legit warnings insofar as those functions behave exactly as documented. 

So our options are:

1. Zero-initialize the buffers before formatting the string
2. Use `StringCchPrintf` versions, which you don't like.

I'm guessing that `1` is the best.

**Comment by randomascii** [comment]
I vote for option 0, do nothing. If swprintf_s and _vsnwprintf_s aren't secure then what the *#@#$ is the point. I'm pretty sure that is an annotation error or a /analyze error. We should file a bug, suppress the warning, and move on.

Unless you explicitly overrule their behavior then they will always null-terminate. If that is not the case then I'm going to quit programming and become a barista.

**Comment by randomascii** [comment]
For the truncation warnings I have an easy fix. However the easy fix requires VC++ 2015. And, the warnings are only marginally real so there is no hurry until we upgrade.

The fix is to cast to ptrdiff_t or some-such instead of to int, and then print with %zd.

**Comment by ariccio** [comment]
> I vote for option 0, do nothing. If swprintf_s and _vsnwprintf_s aren't secure then what the *#@#$ is the point. I'm pretty sure that is an annotation error or a /analyze error. We should file a bug, suppress the warning, and move on.

hahaha....


There is some evidence that it's an annotation error. Replacing the template-parameter-inferenced versions with the versions that explicitly require a buffer size removes the warnings. 

I think it comes down to:
1. A difference in the templated version's annotations (vs the non-templated version)
2. The (annoying) fact that annotations don't propagate up or down the call tree.

The declaration/definition of the templated version of `_vsnwprintf_s` is:

```
__DEFINE_CPP_OVERLOAD_SECURE_FUNC_0_3(
    _Success_(return >= 0)
    int, _vsnwprintf_s,
    _Post_z_                      wchar_t,        _Buffer,
    _In_                          size_t,         _BufferCount,
    _In_z_ _Printf_format_string_ wchar_t const*, _Format,
                                  va_list,        _ArgList
    )
```

...where `__DEFINE_CPP_OVERLOAD_SECURE_FUNC_0_3` is:

```
        #define __DEFINE_CPP_OVERLOAD_SECURE_FUNC_0_3(_ReturnType, _FuncName, _DstType, _Dst, _TType1, _TArg1, _TType2, _TArg2, _TType3, _TArg3) \
            extern "C++"                                                                                                                         \
            {                                                                                                                                    \
                template <size_t _Size>                                                                                                          \
                inline                                                                                                                           \
                _ReturnType __CRTDECL _FuncName(_DstType (&_Dst)[_Size], _TType1 _TArg1, _TType2 _TArg2, _TType3 _TArg3) _CRT_SECURE_CPP_NOTHROW \
                {                                                                                                                                \
                    return _FuncName(_Dst, _Size, _TArg1, _TArg2, _TArg3);                                                                       \
                }                                                                                                                                \
            }
```

...which becomes (manually expanded):

```
                template <size_t _Size>
                inline
                _Success_(return >= 0) int __cdecl _vsnwprintf_s(_Post_z_ wchar_t (&_Buffer)[_Size], _In_ size_t _BufferCount, _In_z_ _Printf_format_string_ wchar_t const* _Format, va_list _ArgList) throw()
                {
                    return _vsnwprintf_s(_Buffer, _Size, _BufferCount, _Format, _ArgList);
                }
```

...and the declaration/definition of the non-templated `_vsnwprintf_s` is:

```
_Success_(return >= 0)
_Check_return_opt_
_CRT_STDIO_INLINE int __CRTDECL _vsnwprintf_s(
    _Out_writes_z_(_BufferCount)  wchar_t*       const _Buffer,
    _In_                          size_t         const _BufferCount,
    _In_                          size_t         const _MaxCount,
    _In_z_ _Printf_format_string_ wchar_t const* const _Format,
                                  va_list              _ArgList
    )
#if defined _NO_CRT_STDIO_INLINE
;
#else
{
    return _vsnwprintf_s_l(_Buffer, _BufferCount, _MaxCount, _Format, NULL, _ArgList);
}
#endif
```

Notice that the templated version annotates `_Buffer` with `_Post_z_`, while the non-templated version uses `_Out_writes_z_(_BufferCount)`.

But after smashing my head against the documentation for a while, I think [what it says](https://msdn.microsoft.com/en-us/library/d3xd30zz.aspx) is that, if the invalid parameter handler doesn't force a crash, the buffer **is guaranteed** to be null terminated.

> If the storage required to store the data and a terminating null exceeds sizeOfBuffer, the invalid parameter handler is invoked, as described in Parameter Validation, unless count is _TRUNCATE, in which case as much of the string as will fit in buffer is written and -1 returned. If execution continues after the invalid parameter handler, these functions set buffer to an empty string, set errno to ERANGE, and return -1.

Uggh, this documentation is going to kill me.

As best as I can tell (I stopped myself halfway down the rabbit hole of `"C:\Program Files (x86)\Windows Kits\10\Source\10.0.10240.0\ucrt\inc\corecrt_internal_stdio_output.h"`), the buffer ****IS**** always actually null terminated.


So, I think we should file a bug?

**Comment by randomascii** [comment]
We should file a bug. I gave a talk about the safer CRT when I worked at Microsoft (I can't find the slides online) and I said that the _s functions always null-terminated and the VC++ guys reviewed my slides and said "looks good".

**Comment by ariccio** [comment]
> We should file a bug. I gave a talk about the safer CRT when I worked at Microsoft


 `*`*snickers*`*` that's a pretty good reason.

**Comment by ariccio** [comment]
So, it looks like you already filed a bug:

[VS 2015 /analyze warns that vsprintf_s doesn't null-terminate buffers - by Bruce Dawson2](https://connect.microsoft.com/VisualStudio/Feedback/Details/1570481)

**Comment by randomascii** [comment]
I though the bug sounded familiar. They've made the same mistake on other functions in the past so it's hard to keep track.

Maybe add a comment on the bug with either a link to or a copy of your analysis of the annotations. And if other functions are affected, be sure to list them as well.

**Comment by ariccio** [comment]
It's interesting that they made the same mistake - I remember a @StephanTLavavej presentation on STL implementation, where he mentioned that Microsoft has a test for basically every STL bug reported. I guess there's nothing like that for `/analyze`.

Fun fact: The context for that mention was a workaround for some quirky iterator operator overloading. The code looked like:
```
for (; _First1 != _Last1 && _First2 != _Last2; ++_First1, (void)++_First2)
```

...where someone overloaded `operator,` broken the STL. The `(void)` cast fixes the precedence.

**Comment by ariccio** [comment]
[51 minutes in, if you're curious](http://www.youtube.com/watch?v=dTeKf5Oek2c)

**Comment by StephanTLavavej** [comment]
We have regression tests for most STL bugs, some of which attempt to exhaustively check for problems of a particular nature. (The comma one is exhaustive.) Our tests are run with over a dozen switch combinations, including /analyze. This ensures that the STL is /W4 /analyze clean. We are especially careful about this because we have so much header-only code, and it stresses the compiler. Other libraries, including the CRT, have less paranoid tests, because they have been burned by warnings and compiler bugs less often. (As an aside, the STL has virtually no analysis annotations, but /analyze can still complain about stuff.) I have a test that verifies that all CRT/STL headers can be included under /W4 /analyze, but it doesn't check usage.

**Comment by randomascii** [comment]
Good to hear Stephan.

It would be nice if the CRT headers were tested a bit more thoroughly, at least to make sure that the _s functions don't give warnings suggesting that they aren't secure. Maybe this incident (not the first of its type!) will encourage them to do so.

**Comment by ariccio** [comment]
My gut reaction is that annotating all of the `_s` functions with `_Always_(_Post_z_)` (vs plain old `_Post_z_`, as is currently) would fix the problem. 

I'm pretty sure that the StrSafe functions are annotated this way already - the only issue I can imagine is that it might interact poorly when someone passes a zero-length buffer (or some other nonsensical address), `/analyze` might thus assume it's suddenly a valid string. That shouldn't be a problem for the templated parameter-inferencing functions.

**Comment by randomascii** [comment]
You should add suggestions on how to annotate the _s functions to the connect bug rather than to this one.

I'll probably upgrade to VS 2015 soon - as soon as the utility PR is landed - and then close this.

**Comment by StephanTLavavej** [comment]
I've pinged James, our CRT maintainer.

**Comment by ariccio** [comment]
Ok, I've added suggestions to the connect bug.

**Comment by randomascii** [comment]
With VS 2015 RTM, and a fix to use %zd to resolve some VS 2015 warnings, the /analyze build gives five warnings:

1>  UIforETW.vcxproj -> C:\github\UIforETW\UIforETW\..\bin\UIforETW_devdeb32.exe
utility.cpp(418): warning C6054: String 'buffer' might not be zero-terminated.
about.h(33): warning C6054: String 'buffer' might not be zero-terminated.
uiforetwdlg.cpp(331): error C6271: Extra argument passed to 'outputPrintf':  _Param_(2) is not used by the format string.
uiforetwdlg.cpp(349): error C6271: Extra argument passed to 'outputPrintf':  _Param_(2) is not used by the format string.
workingset.cpp(147): warning C6054: String 'process' might not be zero-terminated.

They are all spurious. The C6054 warnings are a buggy annotation (previously discussed) and the C6271 warnings are because /analyze doesn't know about %zu. I'm going to try VS 2015 Update 1 RC to see if it is any better.

**Comment by randomascii** [comment]
After installing VS 2015 Update 1 RC I hit worse errors when running code analysis:

1>  UIforETW.vcxproj -> C:\github\UIforETW\UIforETW\..\bin\UIforETW_devdeb32.exe
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018: The "NativeCodeAnalysis" task failed unexpectedly.
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018: Microsoft.VisualStudio.CodeAnalysis.AnalysisResults.AnalysisResultException: CA0001 : An unknown error occurred while running Code Analysis. ---> System.FormatException: Input string was not in a correct format.
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018:    at System.Number.StringToNumber(String str, NumberStyles options, NumberBuffer& number, NumberFormatInfo info, Boolean parseDecimal)
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018:    at System.Number.ParseInt32(String s, NumberStyles style, NumberFormatInfo info)
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018:    at System.Xml.XmlReader.ReadElementContentAsInt()
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018:    at Microsoft.VisualStudio.CodeAnalysis.AnalysisResults.AnalysisResultReader.ReadKeyEvent()
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018:    at Microsoft.VisualStudio.CodeAnalysis.AnalysisResults.AnalysisResultReader.ReadSourceLocation()
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018:    at Microsoft.VisualStudio.CodeAnalysis.AnalysisResults.AnalysisResultReader.ReadSourceTrace()
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018:    at Microsoft.VisualStudio.CodeAnalysis.AnalysisResults.AnalysisResultReader.ReadAnalysisMessage(AnalysisUnit parentUnit)
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018:    at Microsoft.VisualStudio.CodeAnalysis.AnalysisResults.AnalysisResultReader.ReadAnalysisResult()
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018:    at Microsoft.VisualStudio.CodeAnalysis.AnalysisResults.AnalysisResult.LoadFromFile(String filePath, AnalysisResultReaderOptions options)
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018:    at Microsoft.Build.Tasks.NativeCodeAnalysis.Execute()
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018:    --- End of inner exception stack trace ---
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018:    at Microsoft.Build.Tasks.NativeCodeAnalysis.Execute()
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()

There are also a series of 39 extremely verbose warnings - here's one of them:
Severity	Code	Description	Project	File	Line	Suppression State
Warning		The element 'PropertyGroup' in namespace 'http://schemas.microsoft.com/developer/msbuild/2003' has invalid child element 'CodeAnalysisStaticAnalysisDirectory' in namespace 'http://schemas.microsoft.com/developer/msbuild/2003'. List of possible elements expected: 'Property, VisualStudioVersion, MinimumVisualStudioVersion, AdditionalFileItemNames, AllowUnsafeBlocks, AppConfigForCompiler, ApplicationIcon, ApplicationRevision, ApplicationVersion, AppDesignerFolder, AspNetConfiguration, AssemblyKeyContainerName, AssemblyKeyProviderName, AssemblyName, AssemblyOriginatorKeyFile, AssemblyOriginatorKeyFileType, AssemblyOriginatorKeyMode, AssemblyType, AutoGenerateBindingRedirects, AutorunEnabled, BaseAddress, BootstrapperComponentsLocation, BootstrapperComponentsUrl, BootstrapperEnabled, CharacterSet, CheckForOverflowUnderflow, CLRSupport, UseDebugLibraries, CodePage, Configuration, ConfigurationName, ConfigurationOverrideFile, CreateDesktopShortcut, CreateWebPageOnPublish, CurrentSolutionConfigurationContents, DebugSecurityZoneURL, DebugSymbols, DebugType, DefaultClientScript, DefaultHTMLPageLayout, DefaultTargetSchema, DefineConstants, DefineDebug, DefineTrace, DelaySign, DisableLangXtns, DisallowUrlActivation, CodeAnalysisAdditionalOptions, CodeAnalysisApplyLogFileXsl, ....	Miscellaneous Files	C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets	25

**Comment by ariccio** [comment]
Comment nitpick: It's much easier to read logs in comments when they're `formatted` ;)

**Comment by ariccio** [comment]
```
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018: Microsoft.VisualStudio.CodeAnalysis.AnalysisResults.AnalysisResultException: CA0001 : An unknown error occurred while running Code Analysis. ---> System.FormatException: Input string was not in a correct format.
 1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018: at System.Number.StringToNumber(String str, NumberStyles options, NumberBuffer& number, NumberFormatInfo info, Boolean parseDecimal)
 1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018: at System.Number.ParseInt32(String s, NumberStyles style, NumberFormatInfo info)
 1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(377,5): error MSB4018: at System.Xml.XmlReader.ReadElementContentAsInt()
```

Looks like it tried to parse non-numerical text as a number.
It'd be *awfully nice* if it showed the offending text, and maybe even the path to the XML file... If you can find the code analysis result file, then you might be able to spot it. *Probably* not, but then again, it may be obvious.

**Comment by ariccio** [comment]
Scratch that last bit `Microsoft.CodeAnalysis.targets(377,5)` *is* the file and line number.

**Comment by randomascii** [comment]
Well, they give that file/line number for every line in the call stack, so that's not very useful. This is what that line looks like:

<NativeCodeAnalysis Condition="'$(Language)'=='C++' and '$(RunCodeAnalysisOnThisProject)'=='true'"

Replacing 'true' with '1' seemed to get past this error, but /analyze still didn't run.

I've committed the change to VS 2015, purely to simplify reporting this bug. I will test on a VS 2015 Professional installation tomorrow. Regular compiling works fine, however we should avoid adding any VS 2015 specific features until this is resolved, because not having access to /analyze makes me nervous.

The bug is https://connect.microsoft.com/VisualStudio/feedback/details/1971490

**Comment by randomascii** [comment]
Thanks for the suggestion on formatting. I edited my comment and it looks better now.

**Comment by randomascii** [comment]
The /analyze bugs should be fixed in the final VS 2015 Update 1 release. This should also resolve all or most of the false-positive warnings.

One bug was found in the VC++ 2015 version of the code. It's fixed by change 55d6b3a9. The problem happens if you call GetLastError() too far away from the error it is trying to capture. The check-in comment has the details but watch out for this one, it's nasty, and the more you depend on GetLastError() the more vulnerable you are to it.

**Comment by randomascii** [comment]
Microsoft just confirmed that the C6054 non-truncated warnings in http://connect.microsoft.com/VisualStudio/feedback/details/1570481/vs-2015-analyze-warns-that-vsprintf-s-doesnt-null-terminate-buffers are bogus and will be fixed in a later update of the SDK.
